src/nsrt_learning/strips_learning/gen_to_spec_learner.py

```diff
@@ -70,23 +70,65 @@ def _learn(self) -> List[PartialNSRTAndDatastore]:
         self._assert_all_data_in_exactly_one_datastore(
             list(param_opt_to_general_pnad.values()))
 
-        # Pass over the demonstrations multiple times. Each time, backchain
-        # to learn PNADs. Repeat until a fixed point is reached.
-        nec_pnad_set_changed = True
-        while nec_pnad_set_changed:
-            # Before each pass, clear the poss_keep_effects and
-            # seg_to_keep_effects_sub of all the PNADs. We do this because
-            # we only want the poss_keep_effects of the final pass, where
-            # the PNADs did not change.
-            for pnads in param_opt_to_nec_pnads.values():
-                for pnad in pnads:
-                    pnad.poss_keep_effects.clear()
-                    pnad.seg_to_keep_effects_sub.clear()
-            # Run one pass of backchaining.
-            nec_pnad_set_changed = self._backchain_one_pass(
-                param_opt_to_nec_pnads, param_opt_to_general_pnad)
-
-        final_pnads = self._finish_learning(param_opt_to_nec_pnads)
+        curr_finalized_pnads: Set[PartialNSRTAndDatastore] = set()
+        finalized_pnads_converged = False
+
+        while not finalized_pnads_converged:
+            # Pass over the demonstrations multiple times. Each time, backchain
+            # to learn PNADs. Repeat until a fixed point is reached.
+            nec_pnad_set_changed = True
+            while nec_pnad_set_changed:
+                # Before each pass, clear the poss_keep_effects
+                # of all the PNADs. We do this because we only want the
+                # poss_keep_effects of the final pass, where the PNADs did
+                # not change. However, we cannot simply clear the the
+                # sub_dict because some of these substitutions might be
```

Comment:
can you clarify which sub_dict you're talking about? Is it pnad.seg_to_keep_effects_sub?